File: problems/3160_find_the_number_of_distinct_colors_among_the_balls/Solution.java
import java.util.*;

class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        Map<Integer,Integer> ballColor = new HashMap<>(), colorCount = new HashMap<>();
        int[] res = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int ball = queries[i][0], color = queries[i][1];
            if (ballColor.containsKey(ball)) {
                int old = ballColor.get(ball);
                colorCount.merge(old, -1, Integer::sum);
                if (colorCount.get(old) == 0) colorCount.remove(old);
            }
            ballColor.put(ball, color);
            colorCount.merge(color, 1, Integer::sum);
            res[i] = colorCount.size();
        }
        return res;
    }
}

File: problems/3160_find_the_number_of_distinct_colors_among_the_balls/solution.py
from typing import List


class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        ball_color = {}
        color_count = {}
        res = []
        for ball, color in queries:
            if ball in ball_color:
                old = ball_color[ball]
                color_count[old] -= 1
                if color_count[old] == 0:
                    del color_count[old]
            ball_color[ball] = color
            color_count[color] = color_count.get(color, 0) + 1
            res.append(len(color_count))
        return res
